vscode-dotnet-runtime-library/src/Acquisition/InstallTracker.ts

```diff
@@ -0,0 +1,302 @@
+/*---------------------------------------------------------------------------------------------
+*  Licensed to the .NET Foundation under one or more agreements.
+*  The .NET Foundation licenses this file to you under the MIT license.
+*--------------------------------------------------------------------------------------------*/
+import * as fs from 'fs';
+import * as lockfile from 'proper-lockfile';
+import * as path from 'path';
+import { IAcquisitionWorkerContext } from './IAcquisitionWorkerContext';
+import {
+    DotnetAcquisitionStatusResolved,
+    DotnetLockAttemptingAcquireEvent,
+    DotnetLockErrorEvent,
+    DotnetPreinstallDetected,
+    DotnetPreinstallDetectionError,
+    DuplicateInstallDetected,
+    NoMatchingInstallToStopTracking
+} from '../EventStream/EventStreamEvents';
+import {
+    DotnetInstall,
+    GetDotnetInstallInfo,
+    InstallToStrings,
+    IsEquivalentInstallation,
+    IsEquivalentInstallationFile
+} from './DotnetInstall';
+import { getVersionFromLegacyInstallKey, getAssumedInstallInfo } from '../Utils/InstallKeyUtilities';
+import { InstallRecord, InstallRecordOrStr } from './InstallRecord';
+import { DotnetCoreAcquisitionWorker } from './DotnetCoreAcquisitionWorker';
+import { error } from 'console';
+/* tslint:disable:no-any */
+
+
+interface InProgressInstall
+{
+    dotnetInstall: DotnetInstall;
+    // The string is the path of the install once completed.
+    installingPromise: Promise<string>;
+}
+
+
+export class InstallTracker
+{
+    protected inProgressInstalls: Set<InProgressInstall> = new Set<InProgressInstall>();
+    private readonly installingVersionsKey = 'installing';
+    private readonly installedVersionsKey = 'installed';
+
+    public constructor(protected readonly context : IAcquisitionWorkerContext)
+    {
+
+    }
+
+    protected executeWithLock = async <A extends any[], R>(alreadyHoldingLock : boolean, f: (...args: A) => R, ...args: A): Promise<R> =>
+    {
+
+        const trackingLock = 'tracking.lock';
+        const lockPath = path.join(__dirname, trackingLock);
```

Comment:
I get a little weirded out when writing files to the extension install directory - should we be using the state directory here instead?

Reply:
Thank you for taking a look at the code. Sorry it is such a big change!

I'm not sure what you mean by the state directory? (If you mean the extension state, this needs to be an actual file on disk.)

I totally understand your hesitation, with webpack and such this directory can change, and it is volatile. This file is only a temporary one used by the local code however, and `__dirname` is the safest way I know of to get a non-hardcoded directory that we manage.

Reply:
I'm going to get this merged for now since it looks like you approved the broad swath of changes, but we can definitely keep following up on this one!